include/alpaka/dev/DevSycl.hpp

```diff
@@ -0,0 +1,220 @@
+            ~DevSycl() = default;
+
+        public:
+            cl::sycl::device m_Device;
+            cl::sycl::queue m_Queue;
+        };
+    }
+
+    namespace dev
+    {
+        namespace traits
+        {
+            //#############################################################################
+            //! The SYCL device name get trait specialization.
+            template<>
+            struct GetName<
+                dev::DevSycl>
+            {
+                //-----------------------------------------------------------------------------
+                ALPAKA_FN_HOST static auto getName(
+                    dev::DevSycl const & dev)
+                -> std::string
+                {
+                    // get_info returns std::string in this case
+                    return dev.m_Device.get_info<
+                        cl::sycl::info::device::name>();
+                }
+            };
+
+            //#############################################################################
+            //! The SYCL device available memory get trait specialization.
+            template<>
+            struct GetMemBytes<
+                dev::DevSycl>
+            {
+                //-----------------------------------------------------------------------------
+                ALPAKA_FN_HOST static auto getMemBytes(
+                    dev::DevSycl const & dev)
+                -> std::size_t
+                {
+                    // get_info returns cl_ulong in this case
+                    return dev.m_Device.get_info<
+                        cl::sycl::info::device::global_mem_size>();
+                }
+            };
+
+            //#############################################################################
+            //! The SYCL device free memory get trait specialization. Note that
+            //! this function will always return 0 as there is no way in SYCL
+            //! or OpenCL to query free memory.
+            template<>
+            struct GetFreeMemBytes<
+                dev::DevSycl>
+            {
+                //-----------------------------------------------------------------------------
+                ALPAKA_FN_HOST static auto getFreeMemBytes(
+                    dev::DevSycl const & dev)
+                -> std::size_t
+                {
+                    // FIXME: There is no way in either SYCL or OpenCL to
+                    // query free memory.
+                    throw std::runtime_error{"Querying free device memory not supported on SYCL platforms"};
+                }
+            };
+
+            //#############################################################################
+            //! The SYCL device reset trait specialization. Note that this
+            //! function won't actually do anything. If you need to reset your
+            //! SYCL device its destructor must be called.
```

Comment:
We should think about doing this (holding the device as shared_ptr and calling reset and reinitalize it)

Reply:
Do we need a `shared_ptr` for that? Since `cl::sycl::device` is a class member wouldn't be a `unique_ptr` sufficient?

Reply:
The DevXxx classes have the same handle semantics as the basic CUDA device (primitive integer). So they can be copied which a `unique_ptr` would prevent.
I am not sure at all if the current DevSycl would work at all and what happens if the members are copied/destructed.

Reply:
I've given this more thought and I believe there is actually no equivalent to `cudaDeviceReset` in SYCL or OpenCL. Both specifications don't mention anything similar, especially not with regard to a single device. Even if we destruct a SYCL device we have no guarantee that this will actually clear the device state in a similar fashion to `cudaDeviceReset`.

I assume that destroying the entire SYCL/OpenCL context could work somehow, but this would affect all devices and introduce a lot of overhead.